src/library/export/coverartcopyworker.cpp

```diff
@@ -0,0 +1,98 @@
+#include "library/export/coverartcopyworker.h"
+
+#include <QDebug>
+#include <QFileInfo>
+#include <QMessageBox>
+
+#include "util/fileaccess.h"
+#include "util/imagefiledata.h"
+#include "util/safelywritablefile.h"
+
+void CoverArtCopyWorker::run() {
+    auto selectedCover = mixxx::FileAccess(mixxx::FileInfo(m_selectedCoverArtFilePath));
+    ImageFileData imageFileData = ImageFileData::fromFilePath(m_selectedCoverArtFilePath);
+    if (imageFileData.isNull()) {
+        // TODO(rryan): feedback
+        return;
+    }
+
+    m_coverInfo.type = CoverInfo::FILE;
+    m_coverInfo.source = CoverInfo::USER_SELECTED;
+    m_coverInfo.coverLocation = m_selectedCoverArtFilePath;
+    m_coverInfo.setImage(imageFileData);
+
+    if (QFileInfo(m_oldCoverArtFilePath).canonicalPath() ==
+            QFileInfo(m_selectedCoverArtFilePath).canonicalPath()) {
+        qDebug() << "Track and selected cover art are in the same path:"
+                 << QFileInfo(m_selectedCoverArtFilePath).canonicalPath()
+                 << "Cover art updated without copying";
+        emit updateCoverArt(m_coverInfo);
+        return;
+    }
+
+    copyFile(m_selectedCoverArtFilePath, m_oldCoverArtFilePath);
+}
+
+void CoverArtCopyWorker::copyFile(
+        const QString& m_selectedCoverArtFilePath,
+        const QString& m_oldCoverArtFilePath) {
+    QFileInfo coverArtPathFileInfo(m_oldCoverArtFilePath);
+    ImageFileData imageFileData = ImageFileData::fromFilePath(m_selectedCoverArtFilePath);
+
+    if (coverArtPathFileInfo.exists()) {
+        switch (makeOverwriteRequest(m_oldCoverArtFilePath)) {
+        case OverwriteAnswer::Cancel:
+            return;
+        case OverwriteAnswer::Overwrite:
+            break;
+        }
+
+        mixxx::SafelyWritableFile safelyWritableFile(m_oldCoverArtFilePath,
+                mixxx::SafelyWritableFile::SafetyMode::Replace);
+
+        DEBUG_ASSERT(!safelyWritableFile.fileName().isEmpty());
+        if (imageFileData.saveFile(safelyWritableFile.fileName())) {
+            qDebug() << "Cover art"
+                     << m_oldCoverArtFilePath
+                     << "copied successfully";
+            safelyWritableFile.commit();
+        } else {
+            qWarning() << "Error while copying the cover art to" << safelyWritableFile.fileName();
+        }
+    } else {
+        if (imageFileData.saveFile(m_oldCoverArtFilePath)) {
+            qDebug() << "Cover art"
+                     << m_oldCoverArtFilePath
+                     << "copied successfully";
+        } else {
+            qWarning() << "Error while copying the cover art to" << m_oldCoverArtFilePath;
```

Comment:
all the error conditions should be also emitted. Not sure if we should go for an extra signal or an error enum in the existing signal.

Reply:
IMHO Enums would make this class a bit complicated, we can just emit the signal via QString and inform user about the copying cover art failed. Just in case if they might want to check the situation.

The info message can be something like,
`Error while copying the cover art to location:`
`Error while copying the _temp cover art to location.,

If that happens we can get instant feedback from the user and then fix it.

Reply:
Logging the warning is probably sufficient. Just signal the caller whether copying was not successful or not.

Reply:
Oh, I just noticed that you have already implemented the message box. Nice, no need for a change.